src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp

```diff
@@ -124,152 +72,137 @@ PiecewisePolynomial<MaxDeriv>::func_and_derivs(const double t) const {
   // function. Retain support for unpacking data written by previous versions
   // whenever possible. See `Domain` docs for details.
 
-  // Here is where we must retain support for older versions
-  if (p.isUnpacking()) {
-    // For versions 0 and 1, we stored the data first, then expiration time,
-    // then possibly the size
-    if (version <= 1) {
-      if (version == 0) {
-        // Version 0 had a std::vector rather than a std::list
-        std::vector<FunctionOfTimeHelpers::StoredInfo<MaxDeriv + 1>>
-            pupped_info{};
-        p | pupped_info;
-        for (auto& pupped_deriv_info : pupped_info) {
-          deriv_info_at_update_times_.emplace_back(
-              std::move(pupped_deriv_info));
-        }
-      } else {
-        // Version 1 had a std::deque rather than a std::list
-        std::deque<FunctionOfTimeHelpers::StoredInfo<MaxDeriv + 1>>
-            pupped_info{};
-        p | pupped_info;
-        for (auto& pupped_deriv_info : pupped_info) {
-          deriv_info_at_update_times_.emplace_back(
-              std::move(pupped_deriv_info));
-        }
-      }
+  if (version < 3) {
+    unpack_old_version(p, version);
+    return;
+  }
+
+  p | deriv_info_at_update_times_;
+}
 
-      // Same for v0 and v1
-      p | expiration_time_;
+namespace {
+template <size_t MaxDerivPlusOne>
+struct LegacyStoredInfo {
+  double time{std::numeric_limits<double>::signaling_NaN()};
+  std::array<DataVector, MaxDerivPlusOne> stored_quantities;
 
-      if (version == 0) {
-        deriv_info_size_.store(deriv_info_at_update_times_.size(),
-                               std::memory_order_release);
-      } else {
-        p | deriv_info_size_;
-      }
-    } else if (version >= 2) {
-      // However, for v2+, we store expiration time, size, then data for
-      // thread-safety reasons
-      p | expiration_time_;
-      size_t size = 0;
-      p | size;
-      deriv_info_size_.store(size, std::memory_order_release);
-      deriv_info_at_update_times_.clear();
-      deriv_info_at_update_times_.resize(
-          deriv_info_size_.load(std::memory_order_acquire));
-      // Using range-based loop here is ok because we won't be updating while
-      // packing/unpacking
-      for (auto& deriv_info : deriv_info_at_update_times_) {
-        p | deriv_info;
-      }
+  void pup(PUP::er& p) {
+    p | time;
+    p | stored_quantities;
+  }
+};
+}  // namespace
+
+template <size_t MaxDeriv>
+void PiecewisePolynomial<MaxDeriv>::unpack_old_version(PUP::er& p,
+                                                       const size_t version) {
+  ASSERT(p.isUnpacking(), "Can't serialize old version");
+
+  std::vector<LegacyStoredInfo<MaxDeriv + 1>> info_vector{};
+  double expiration_time{};
+
+  // For versions 0 and 1, we stored the data first, then expiration time,
+  // then possibly the size
+  if (version <= 1) {
+    if (version == 0) {
+      // Version 0 had a std::vector rather than a std::list
```

Comment:
Remove `rather than a std::list` (this function doesn't know anything about lists). Same in QuatFoT